src/Microsoft.Extensions.DependencyInjection.Abstractions/ServiceDescriptor.cs

```diff
@@ -95,46 +22,16 @@ private ServiceDescriptor(Type serviceType, ServiceLifetime lifetime)
         /// <inheritdoc />
         public Type ServiceType { get; }
 
-        /// <inheritdoc />
-        public Type ImplementationType { get; }
-
-        /// <inheritdoc />
-        public object ImplementationInstance { get; }
-
-        /// <inheritdoc />
-        public Func<IServiceProvider, object> ImplementationFactory { get; }
-
-        internal Type GetImplementationType()
-        {
-            if (ImplementationType != null)
-            {
-                return ImplementationType;
-            }
-            else if (ImplementationInstance != null)
-            {
-                return ImplementationInstance.GetType();
-            }
-            else if (ImplementationFactory != null)
-            {
-                var typeArguments = ImplementationFactory.GetType().GenericTypeArguments;
-
-                Debug.Assert(typeArguments.Length == 2);
-
-                return typeArguments[1];
-            }
-
-            Debug.Assert(false, "ImplementationType, ImplementationInstance or ImplementationFactory must be non null");
-            return null;
-        }
+        internal abstract Type GetImplementationType();
```

Comment:
I guess we can leave this internal until we decide to open up `ServiceDescriptor`